packages/@react-spectrum/tabs/src/Tabs.tsx

```diff
@@ -121,40 +177,228 @@ export function Tab<T>(props: TabProps<T>) {
+  wrapperRef: MutableRefObject<HTMLDivElement>,
+  collapse?: boolean
+}
+
+const CollapsibleTabList = React.forwardRef(function <T> (props: CollapsibleTabListProps<T>, ref: MutableRefObject<HTMLDivElement>) {
+  let {
+    tabListProps,
+    density,
+    isQuiet,
+    isDisabled,
+    state,
+    selectedTab,
+    wrapperRef,
+    collapse
+  } = props;
+
+  let tablist;
+  if (collapse) {
+    tablist = <TabPicker {...props} />;
+  } else {
+    tablist = (
+      <TabList
+        {...tabListProps}
+        density={density}
+        isQuiet={isQuiet}
+        isDisabled={isDisabled}
+        state={state}
+        selectedTab={selectedTab}
+        ref={ref}
+        orientation="horizontal" />
+    );
+  }
+
+  return (
+    <div
+      ref={wrapperRef}
+      className={classNames(
+        styles,
+        'spectrum-Tabs--collapsible'
+      )}>
+      {tablist}
+    </div>
+  );
+});
+
+interface TabListProps<T> {
+  isQuiet?: boolean,
+  density?: 'compact',
+  isDisabled?: boolean,
+  orientation?: Orientation,
+  state: SingleSelectListState<T>,
+  selectedTab: HTMLElement
+}
+
+const TabList = React.forwardRef(function <T> (props: TabListProps<T>, ref: MutableRefObject<HTMLDivElement>) {
+  let {
+    isQuiet,
+    density,
+    state,
+    isDisabled,
+    orientation,
+    selectedTab,
+    ...otherProps
+  } = props;
+
+  return (
+    <div
+      {...otherProps}
+      ref={ref}
+      className={classNames(
+        styles,
+        'spectrum-Tabs',
+        `spectrum-Tabs--${orientation}`,
+        {
+          'spectrum-Tabs--quiet': isQuiet,
+          [`spectrum-Tabs--${density}`]: density
+        },
+        'spectrum-Tabs-container'
+      )}>
+      {[...state.collection].map((item) => (
+        <Tab key={item.key} item={item} state={state} isDisabled={isDisabled} orientation={orientation} />
+      ))}
+      <TabLine orientation={orientation} selectedTab={selectedTab} />
+    </div>
+  );
+});
+
+interface TabPickerProps<T> extends SpectrumPickerProps<T> {
+  density?: 'compact',
+  state: SingleSelectListState<T>
+}
+
+function TabPicker<T>(props: TabPickerProps<T>) {
+  let {
+    isDisabled,
+    isQuiet,
+    state,
+    'aria-labelledby': ariaLabeledBy,
+    'aria-label': ariaLabel,
+    density
+  } = props;
+
+  let ref = useRef();
+  let [pickerNode, setPickerNode] = useState(null);
+
+  useEffect(() => {
+    let node = unwrapDOMRef(ref);
+    setPickerNode(node.current);
+  }, [ref]);
+
+  let items = [...state.collection].map((item) => ({
+    rendered: item.rendered,
+    textValue: item.textValue,
+    key: item.key
+  }));
+
+  let pickerProps = {
+    'aria-labelledby': ariaLabeledBy,
+    'aria-label': ariaLabel
+  };
+
+  // TODO: Figure out if tabListProps should go onto the div here, v2 doesn't do it
+  return (
+    <div
+      className={classNames(
+        styles,
+        'spectrum-Tabs',
+        'spectrum-Tabs--horizontal',
+        'spectrum-Tabs-dropdown',
```

Comment:
interesting, how can this div be both spectrum-Tabs and spectrum-Tabs-dropdown? those two names imply this
```
<div spectrum-Tabs>
  <div spectrum-Tabs-dropdown>
```

and not
```<div spectrum-Tabs spectrum-Tabs-dropdown>```

single dash indicates hierarchy/components
double dash indicates a state for the current component, which is why --horizontal is allowed on spectrum-Tabs because it describes the state of the Tabs
same with quite and density

Reply:
ah, so it should be `spectrum-Tabs--dropdown`?

Reply:
nope, the examples you gave are correct, unless collapsible isn't a state? and container isn't an element?

what I'm trying to say is that dropdown is a component, but if it's a state then maybe it's `spectrum-Tabs--showDropdown`?

Reply:
here's another example (made up, but accurate naming convention)
```
<div spectrum-Dialog spectrum-Dialog--error>
  <div spectrum-Dialog-container>
    <div spectrum-ButtonGroup spectrum-ButtonGroup--horizontal>
      <button spectrum-Dialog-button spectrum-Button spectrum-Button--secondary>
      <button spectrum-Dialog-button spectrum-Button spectrum-Button--primary spectrum-Button--disabled>
```
states/variants are `--` while components are `-`